### PR TITLE
API Move some code to framework to be reusable

### PR DESCRIPTION
**Pull request by GuySartorelli**
This logic is used by `CMSMain` but needs to be callable on any hierarchical model.

See [the corresponding framework PR](https://github.com/silverstripe/silverstripe-framework/pull/11460) to see where this code is going.

See [the overall refactoring PR](https://github.com/silverstripe/silverstripe-cms/pull/3017) (or just look in CMSMain) to see where it's being used and why it needs to be moved. Basically, anything `CMSMain` calls needs to be available on a `DataObject` record which has the `Hierarchy` extension applied.

## Issue
- https://github.com/silverstripe/silverstripe-cms/issues/2947

### CODE REVIEW

File: code/Model/SiteTree.php [hunk 1]
```diff
@@ -114,31 +114,6 @@
...
-        SiteTree::class
-    ];
-
-    /**
-     * Used as a cache for `SiteTree::allowedChildren()`
-     * Drastically reduces admin page load when there are a lot of page types
-     * @var array
-     */
-    protected static $_allowedChildren = [];
-
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php]: Moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 2]
```diff
@@ -158,7 +133,7 @@ class SiteTree extends DataObject implements PermissionProvider, i18nEntityProvi
      * @config
      * @var string
      */
-    private static $default_child = "Page";
+    private static $default_child = Page::class;
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:136]: Unrelated cleanup - this is a class name, not an arbitrary string.

File: code/Model/SiteTree.php [hunk 3]
```diff
@@ -167,16 +142,7 @@ class SiteTree extends DataObject implements PermissionProvider, i18nEntityProvi
...
-    private static $default_classname = "Page";
-
-    /**
-     * The default parent class for this page.
-     * Note: Value might be cached, see {@link $allowed_chilren}.
-     *
-     * @config
-     * @var string
-     */
-    private static $default_parent = null;
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:180]: default_parent moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 4]
```diff
@@ -167,16 +142,7 @@ class SiteTree extends DataObject implements PermissionProvider, i18nEntityProvi
...
+    private static $default_classname = Page::class;
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:145]: Unrelated cleanup - default_classname is a class name, not an arbitrary string.

File: code/Model/SiteTree.php [hunk 5]
```diff
@@ -330,6 +296,8 @@ class SiteTree extends DataObject implements PermissionProvider, i18nEntityProvi
 
     private static $default_sort = "\"Sort\"";
 
+    private static string $sort_field = 'Sort';
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:299]: New config from `Hierarchy` in framework. This allows `CMSMain` and `Hierarchy` both to reason about arbitrary models without requiring a hardcoded specific field name for sorting.

File: code/Model/SiteTree.php [hunk 6]
```diff
@@ -428,29 +394,11 @@ class SiteTree extends DataObject implements PermissionProvider, i18nEntityProvi
...
-     * Description of the class functionality, typically shown to a user
-     * when selecting which page type to create. Translated through {@link provideI18nEntities()}.
-     *
-     * @see SiteTree::classDescription()
-     * @see SiteTree::i18n_classDescription()
-     *
-     * @config
-     * @var string
-     */
-    private static $description = null;
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:443]: Moved to `DataObject` in framework and renamed to a more specific `class_description`

File: code/Model/SiteTree.php [hunk 7]
```diff
@@ -428,29 +394,11 @@ class SiteTree extends DataObject implements PermissionProvider, i18nEntityProvi
...
-     *
-     * @see SiteTree::classDescription()
-     * @see SiteTree::i18n_classDescription()
-     *
-     * @config
-     * @var string
+     * override class_description in subclasses instead.
      */
-    private static $base_description = 'Generic content page';
+    private static string $base_class_description = 'Generic content page';
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:401]: Renamed to match the `description` => `class_description` renaming

File: code/Model/SiteTree.php [hunk 8]
```diff
@@ -891,32 +839,6 @@ protected function onBeforeDuplicate($original, $doWrite)
         $this->Sort = 0;
     }
 
-    /**
-     * Duplicates each child of this node recursively and returns the top-level duplicate node.
-     *
-     * @return static The duplicated object
-     */
-    public function duplicateWithChildren()
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:907]: Moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 9]
```diff
@@ -1095,47 +1017,6 @@ public function can($perm, $member = null, $context = [])
...
-     * - There are no classes defined in {@link $allowed_children}
-     *
-     * @uses SiteTreeExtension->canAddChildren()
-     * @uses canEdit()
-     * @uses $allowed_children
-     *
-     * @param Member|int $member
-     * @return bool True if the current user can add children
-     */
-    public function canAddChildren($member = null)
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:1122]: Moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 10]
```diff
@@ -1305,14 +1186,6 @@ public function canCreate($member = null, $context = [])
             $member = Security::getCurrentUser();
         }
 
-        // Check parent (custom canCreate option for SiteTree)
-        // Block children not allowed for this parent type
-        $parent = isset($context['Parent']) ? $context['Parent'] : null;
-        $strictParentInstance = ($parent && $parent instanceof SiteTree);
-        if ($strictParentInstance && !in_array(static::class, $parent->allowedChildren() ?? [])) {
-            return false;
-        }
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:1322]: Moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 11]
```diff
@@ -1324,14 +1197,16 @@ public function canCreate($member = null, $context = [])
             return true;
         }
 
-        // Fall over to inherited permissions
-        if ($strictParentInstance && $parent->exists()) {
-            return $parent->canAddChildren($member);
-        } else {
+        // Check parent (added to context through CMSMain)
+        $parent = isset($context['Parent']) ? $context['Parent'] : null;
+        if (!$parent?->exists() || !($parent instanceof SiteTree)) {
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:1202]: The `canAddChildren()` check is now done in `Hierarchy` in framework. Instead, here we only really care about whether the record is in this hierarchy and exists.

File: code/Model/SiteTree.php [hunk 12]
```diff
@@ -1324,14 +1197,16 @@ public function canCreate($member = null, $context = [])
...
             // This doesn't necessarily mean we are creating a root page, but that
             // we don't know if there is a parent, so default to this permission
             return SiteConfig::current_site_config()->canCreateTopLevel($member);
         }
+
+        // If we haven't returned by now, resort to edit permissions
+        return $this->canEdit($member);
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:1209]: Gives a clean and sensible fallback in case we haven't returned by here, instead of just returning nothing.

File: code/Model/SiteTree.php [hunk 13]
```diff
@@ -1812,12 +1687,6 @@ protected function onAfterDelete()
         parent::onAfterDelete();
     }
 
-    public function flushCache($persistent = true)
-    {
-        parent::flushCache($persistent);
-        $this->_cache_statusFlags = null;
-    }
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:1827]: No longer needed on this class (but exists on parent class)

File: code/Model/SiteTree.php [hunk 14]
```diff
@@ -1842,41 +1711,6 @@ public function validate()
     {
         $result = parent::validate();
 
-        // Allowed children validation
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:1853]: All this validation was moved to `Hierarchy` in framework since that's where the associated methods and config are.

File: code/Model/SiteTree.php [hunk 15]
```diff
@@ -2786,54 +2620,6 @@ protected function getClassDropdown()
         return $result;
     }
 
-    /**
-     * Returns an array of the class names of classes that are allowed to be children of this class.
-     *
-     * @return string[]
-     */
-    public function allowedChildren()
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:2802]: Moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 16]
```diff
@@ -2869,34 +2655,6 @@ public function creatableChildPages()
         return $children[$this->ID];
     }
 
-    /**
-     * Returns the class name of the default class for children of this page.
-     *
-     * @return string
-     */
-    public function defaultChild()
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:2885]: Moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 17]
```diff
@@ -2869,34 +2655,6 @@ public function creatableChildPages()
...
-        }
-        return null;
-    }
-
-    /**
-     * Returns the class name of the default class for the parent of this page.
-     *
-     * @return string
-     */
-    public function defaultParent()
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:2903]: Moved to `Hierarchy` in framework.

File: code/Model/SiteTree.php [hunk 18]
```diff
@@ -2927,55 +2685,6 @@ public function setMenuTitle($value)
...
-     * Example (simple):
-     *   "deletedonlive" => "Deleted"
-     *
-     * Example (with optional title attribute):
-     *   "deletedonlive" => ['text' => "Deleted", 'title' => 'This page has been deleted']
-     *
-     * @param bool $cached Whether to serve the fields from cache; false regenerate them
-     * @return array
-     */
-    public function getStatusFlags($cached = true)
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:2953]: The method itself is moved to `ModelData` in framework. The `Versioned` extension updates it to include all these version-specific flags (which means all versioned models get it now, not just `SiteTree`)

File: code/Model/SiteTree.php [hunk 19]
```diff
@@ -3244,50 +2953,15 @@ public function getPageIconURL()
...
-        }
-        return $this->config()->get('description');
-    }
-
-    /**
-     * Get localised description for this page
-     *
-     * @return string|null
-     */
-    public function i18n_classDescription()
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php]: Moved to `DataObject` in framework.

File: code/Model/SiteTree.php [hunk 20]
```diff
@@ -3244,50 +2953,15 @@ public function getPageIconURL()
...
-        return null;
-    }
-
-    /**
-     * Overloaded to also provide entities for 'Page' class which is usually located in custom code, hence textcollector
-     * picks it up for the wrong folder.
-     *
-     * @return array
-     */
-    public function provideI18nEntities()
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:3288]: Moved the description entity to `DataObject` in framework. This method is no longer needed in this class.

File: code/Model/VirtualPage.php [hunk 21]
```diff
@@ -179,6 +179,18 @@ public function allowedChildren()
         return [];
     }
 
+    /**
+     * Get the record to check against for allowed children check in validation.
+     */
+    public function getRecordForAllowedChildrenValidation(): SiteTree
```
Comment (ChloeSartorelli) [inline code/Model/VirtualPage.php:185]: See `Hierarchy::updateValidate()` in framework

File: tests/php/Model/SiteTreeTest.php
Comment (ChloeSartorelli) [inline tests/php/Model/SiteTreeTest.php:1]: Tests moved along with the logic they're testing

File: tests/php/Model/SiteTreeTest_ClassCext.php
Comment (ChloeSartorelli) [inline tests/php/Model/SiteTreeTest_ClassCext.php:1]: This class was only ever used in a test that got moved

File: tests/php/Model/VirtualPageTest.php [hunk 22]
```diff
@@ -520,6 +520,8 @@ public function testPageTypeChangePropagatesToLive()
         // but we want to test that it gets copied on class name change instead
         $page->write();
 
+        // We need to flush the static cache so we don't have a stale reference to $page in $virtual->components
+        $virtual->flushCache();
```
Comment (ChloeSartorelli) [inline tests/php/Model/VirtualPageTest.php:524]: This test was really flakey - it relied on the fact that we hadn't fetched the `CopyContentFrom` has_one record yet. Changes in this and the framework PR resulted in it being fetched so we had a stale reference to the `$page` record that got originally saved (with the "published title" text).

We have to clear that out of the static cache here, so that we fetch the record with the "original" text so that can be copied through to our virtual page on write.

File: code/Model/SiteTree.php [hunk 23]
```diff
@@ -3008,6 +2709,7 @@ public function getTreeTitle()
     {
         $children = $this->creatableChildPages();
         $flags = $this->getStatusFlags();
+        $this->extend('updateStatusFlagsForTreeTitle', $flags);
```
Comment (ChloeSartorelli) [inline code/Model/SiteTree.php:2712]: This new extension hook allows fluent to exclude its flags from the site tree list, because they look awful there.
In the next set of PRs after this one, it'll be on `CMSMain` instead, but for now `SiteTree` directly handles how things look in that tree.

**Review by emteknetnz — APPROVED** [review]
